keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+make "fail closed" a safe mode of operation. As a concrete example, only
+allowing CEL expressions that pass compilation and type checking significantly
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare zero trust policies is
+important to compliance. i.e., it should be possible for policies governing 
+resources like namespace and roles to default to the most restrictive state
+when first created.
+
+## Proposal
+
+### Phase 1
+
+Introduce a new "ValidatingAdmissionConfiguration" kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API needs to support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to webhooks MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of details (response message formatting, severity
+levels, failure policies, type safety, ...) that will be discussed in detail
+further on in this proposal.
+
+#### Configurability in the API
+
+How configurability is expressed in the API will have significant influence
+how the API is shaped. We are currently considering two design options:
+
+##### Configurability Option 1: Embed configuration into the object, grouped with match criteria
+
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionConfiguration
+metadata:
+  name: "validate-xyz.example.com"
+configurationSchema:  
+  openAPIV3Schema: ... # the below configuration is validated against this schema
+spec:
+  match:
+  - rules: ... # see ValidatatingWebhookConfiguration rules
+    objectSelectors: ...  # "
+    namespaceSelectors ... # "
+    configuration: {x: 1} # each match may be configured differently
+  - rules: ...
+    objectSelectors: ...
+    namespaceSelectors ...
+    configuration: {x: 2}
+validations:
+  - rule: "self.spec.xyz == configuration.x"
+    # ...other rule related fields here...
+```
+
+Pros:
+
+- Self-contained resource
+- Higher level policy abstractions can be defined out-of-tree and can be
+  translated into this form by controllers
+
+Cons:
+
+- The resource grows by the number of match rules and amounts of configuration data.
+- Need to define how configurations of overlapping matches are interpreted. If
```

Comment:
not exactly related to the con noted here, but this seemed a useful place to mention prior art references for structure of an ordered set of matching rules:
* rbac, priority & fairness, and admission webhook configurations all use ordered lists of rules, and first match allows/matches
* audit configuration also uses an ordered list of rules, but can make a match exclude auditing... this approach could address the "ability to skip specific resource types" mentioned above

Reply:
Thx! Added. We're still nailing down what exact match rules we want to make available.